experimental/incremental/task.go

```diff
@@ -187,116 +196,117 @@ type result struct {
 	done chan struct{}
 }
 
-// run executes a query in the context of some task and writes the result to
-// out.
-func run(caller Task, task *task, q Query[any], wg *semaphore.Weighted, done func()) (async bool) {
-	// Common case for cached values; no need to spawn a separate goroutine.
-	r := task.result.Load()
-	if r != nil && closed(r.done) {
-		done()
-		return false
+// path is a linked list node for tracking cycles in query dependencies.
+type path struct {
+	Query *AnyQuery
+	Prev  *path
+}
+
+// Walk returns an iterator over the linked list.
+func (p *path) Walk() iter.Seq[*path] {
+	return func(yield func(*path) bool) {
+		for node := p; node.Query != nil; node = node.Prev {
+			if !yield(node) {
+				return
+			}
+		}
 	}
+}
 
-	if wg.Acquire(caller.ctx, 1) != nil {
+// run executes a query in the context of some task and writes the result to out.
+func (t *task) run(caller Task, q *AnyQuery, done func(*result)) (async bool) {
```

Comment:
I think this and the one below are confusingly named. `runAny` actually blocks for the result and isn't really async. _This_ is the async version, since it returns immediately and can spawn a goroutine to actually invoke `done`.

Maybe rename this one to `start` and the other to `run`?